src/vs/workbench/parts/terminal/electron-browser/terminalInstance.ts

```diff
@@ -585,16 +589,34 @@ export class TerminalInstance implements ITerminalInstance {
 		if (!this._shellLaunchConfig.executable) {
 			this._configHelper.mergeDefaultShellPathAndArgs(this._shellLaunchConfig);
 		}
-		this._initialCwd = this._getCwd(this._shellLaunchConfig, this._historyService.getLastActiveWorkspaceRoot('file'));
 
-		// Merge process env with the env from config
-		const envFromConfig = { ...process.env };
+		const lastActiveWorkspaceRootUri = this._historyService.getLastActiveWorkspaceRoot('file');
+		const lastActiveWorkspaceRoot = this._workspaceContextService.getWorkspaceFolder(lastActiveWorkspaceRootUri);
+		this._initialCwd = this._getCwd(this._shellLaunchConfig, lastActiveWorkspaceRootUri);
+
+		// Resolve env vars from config and shell
 		const envSettingKey = platform.isWindows ? 'windows' : (platform.isMacintosh ? 'osx' : 'linux');
-		TerminalInstance.mergeEnvironments(envFromConfig, this._configHelper.config.env[envSettingKey]);
+		const envFromConfig = { ...this._configHelper.config.env[envSettingKey] };
+		Object.keys(envFromConfig).forEach((key) => {
```

Comment:
This section's getting a bit crazy, how about we organize it like this:

- `mergeEnvironments` now takes a base environment and an array of secondary environments
- The secondary environments are filtered through a `resolveConfigurationVariables`

I think we could then minimize the number of lines dealing with the env in this function and de-duplicate the `_configurationResolverService` parts. What do you think?

Reply:
I thought exactly that about the `mergeEnvironments`. Wasnt sure what you'd prefer. I'll make the changes

Reply:
One issue with that though is that `mergeEnvironments` is static. Resolving of variables need the `configurationResolverService` and the workspaceRoot

Reply:
If changing the static is too much work then you can pass in _configurationResolveService as a param.